AK/Optional.h

```diff
@@ -185,6 +185,119 @@ class [[nodiscard]] Optional {
     alignas(T) u8 m_storage[sizeof(T)];
     bool m_has_value { false };
 };
+
+template<typename T>
+class [[nodiscard]] Optional<T&> {
+public:
+    using ValueType = T&;
+
+    ALWAYS_INLINE Optional() = default;
+
+    ALWAYS_INLINE Optional(Optional const& other) = delete;
+    ALWAYS_INLINE Optional(Optional&& other)
+        : m_ptr(other.m_ptr)
+    {
+        other.m_ptr = nullptr;
+    }
+
+    ALWAYS_INLINE Optional(T& value)
+        : m_ptr(&value)
+    {
+    }
+
+    ALWAYS_INLINE ~Optional() = default;
+
+    // Avoid the question of deep-copy vs. shallow-copy by simply deleting copy-assign.
+    // Note that the copy-constructor does a shallow copy, just like T&.
+    ALWAYS_INLINE Optional& operator=(Optional const&) = delete;
+    ALWAYS_INLINE Optional& operator=(Optional&& other)
+    {
+        m_ptr = other.m_ptr;
+        other.m_ptr = nullptr;
+        return *this;
+    }
```

Comment:
I know I'm very insisting about this, but please `delete` this as well.
Consider the following:
```c++
int a = 3;
int b = 5;
int& a_r = a;
int& b_r = b;
a_r = std::move(b_r);
```
=> `a: 5, b: 5, a_r: 5, b_r: 5`
```c++
int a = 3;
int b = 5;
Optional<int&> a_r = a;
Optional<int&> b_r = b;
```
=> `a: 3, b: 5, a_r: 5, b_r: {}`
Now, the state of `b_r` doesn't matter because it has just been moved from. What matters is that `a_r` in the first example is deeply assigned to, changing the state of `a`, while in the second example, `a_r` is only shallowly assigned to, not changing `a`. Once again, I think that *if* we were to allow this operation, it should be deep-if-possible. 
[Godbolt Link](https://godbolt.org/z/czh1h1j9a)

Reply:
Throwing this here for the sake of completeness: the semantically equivalent `std::optional<std::reference_wrapper<T>>` also behaves like this PR in that it does shallow copies:
```cpp
int a = 3;
int b = 5;
std::optional<std::reference_wrapper<int>> a_r = a;
std::optional<std::reference_wrapper<int>> b_r = b;
a_r = std::move(b_r);
```
=> `a: 3, b: 5, a_r: 5, b_r: 5`

[Godbolt link](https://godbolt.org/z/8vfoEEPKT)